### PR TITLE
[L0] Enable Disabling of Queue lock during L0 Sync calls

### PR DESCRIPTION
**Pull request by nrspruit**
- Prevent Queue Lock deadlock while L0 queue sync or host sync is in progress.

### CODE REVIEW

File: source/adapters/level_zero/queue.cpp
Comment (pbalcer) [inline source/adapters/level_zero/queue.cpp:1315]: we can only get here if the Queue's refcount reaches 0. Which shouldn't happen while there's a synchronize happening on that queue in parallel.

File: source/adapters/level_zero/queue.cpp
Comment (pbalcer) [inline source/adapters/level_zero/queue.cpp:1441]: dropping the `Mutex` and acquiring the `zeQueueInUseMutex` would have to be atomic for this to be safe. Otherwise, assuming RefCount can be 0 here, a thread calling `urQueueReleaseInternal` could delete the queue object in the unsynchronized gap between these locks.

But, again, I don't believe `zeQueueInUseMutex` is even necessary.

File: source/adapters/level_zero/queue.hpp
Comment (pbalcer) [inline source/adapters/level_zero/queue.hpp:553]: missing empty line :)

File: source/adapters/level_zero/queue.cpp
Comment (pbalcer) [inline source/adapters/level_zero/queue.cpp:1474]: This happens 3 times. Maybe this should be a function in Queue?

File: source/adapters/level_zero/queue.cpp
Comment (nrspruit) [reply source/adapters/level_zero/queue.cpp:1441]: I had added the mutex because originally I was not sure, but let me check again if there is any possibility. If there isn't, then I will remove the mutex all together and instead have the "queue sync" be in a "lockless" mode while awaiting the return from the sync.

File: source/adapters/level_zero/queue.cpp
Comment (nrspruit) [reply source/adapters/level_zero/queue.cpp:1315]: I will recheck to see if there is any possibility of the queue to be deleted while a thread is calling the sync, if not, then I will remove the mutex entirely.

File: source/adapters/level_zero/queue.hpp
Comment (nrspruit) [reply source/adapters/level_zero/queue.hpp:553]: Will fix, weird formatting change.

File: source/adapters/level_zero/queue.cpp
Comment (nrspruit) [reply source/adapters/level_zero/queue.cpp:1456]: I checked the templating and it seems to add to much complexity for this case. I left this out to make the change simpler for now.

**Review by pbalcer — APPROVED** [review]
If we merge this, we should also be able to revert https://github.com/oneapi-src/unified-runtime/pull/1371 (but I'm not sure whether we should given the opt-out `UrL0QueueSyncNonBlocking` flag)

**Comment by kbenzie** [comment]
@pbalcer has approved this but this is still pending review from the `@oneapi-src/unified-runtime-level-zero-write` team. Should @pbalcer (and others?) be in that review team?

**Comment by nrspruit** [comment]
> @pbalcer has approved this but this is still pending review from the `@oneapi-src/unified-runtime-level-zero-write` team. Should @pbalcer (and others?) be in that review team?

@raiyanla is also part of that team so it should have caught his approval, does he need to reapprove?

**Comment by kbenzie** [comment]
> @raiyanla is also part of that team so it should have caught his approval, does he need to reapprove?

Weird, perhaps this was a draft PR when that approval happened? Or GitHub being weird. Either way I think its fine, was just a thought about process moving forwards.

**Review by raiyanla — APPROVED** [review]

**Pull request merged**
